File: public/js/shared/sync-scheduler.js
/**
 * SyncScheduler - Shared sync triggering logic for auto-sync
 * Each module (coach, journal) creates its own instance with its own syncFn.
 * Handles debounced uploads, periodic polling, retry with backoff,
 * and online/visibility event listeners.
 */
import { log as debugLog } from './debug-log.js';
import { showNotification } from './notifications.js';

export class SyncScheduler {
    /**
     * @param {Object} options
     * @param {string} options.name - Module name for logging ('coach', 'journal')
     * @param {Function} options.syncFn - The sync function to call. Returns { success, reason?, error? }
     * @param {Function} options.getIsSyncing - Returns current syncing state
     * @param {Function} options.getHasDirtyData - Returns whether there's dirty data to upload
     * @param {Function} [options.pollCheckFn] - Optional pre-poll check. Return true to proceed with sync.
     * @param {number} [options.uploadDebounceMs=2500] - Debounce delay for upload triggers
     * @param {number} [options.pollIntervalMs=30000] - Polling interval
     * @param {number} [options.baseRetryMs=5000] - Base retry delay
     * @param {number} [options.maxRetryMs=120000] - Max retry delay
     */
    constructor({ name, syncFn, getIsSyncing, getHasDirtyData, pollCheckFn = null,
                  uploadDebounceMs = 2500, pollIntervalMs = 30000,
                  baseRetryMs = 5000, maxRetryMs = 120000 }) {
        this._name = name;
        this._syncFn = syncFn;
        this._getIsSyncing = getIsSyncing;
        this._getHasDirtyData = getHasDirtyData;
        this._pollCheckFn = pollCheckFn;
        this._uploadDebounceMs = uploadDebounceMs;
        this._pollIntervalMs = pollIntervalMs;
        this._baseRetryMs = baseRetryMs;
        this._maxRetryMs = maxRetryMs;

        this._debounceTimer = null;
        this._pollTimer = null;
        this._retryTimer = null;
        this._retryAttempt = 0;
        this._pendingSync = false;
        this._started = false;

        this._onOnline = this._onOnline.bind(this);
        this._onOffline = this._onOffline.bind(this);
        this._onVisibilityChange = this._onVisibilityChange.bind(this);
    }

    /** Set up event listeners and start polling. Call after store init. */
    start() {
        if (this._started) return;
        this._started = true;
        window.addEventListener('online', this._onOnline);
        window.addEventListener('offline', this._onOffline);
        document.addEventListener('visibilitychange', this._onVisibilityChange);
        if (navigator.onLine) {
            this._startPolling();
        }
    }

    /** Teardown: clear all timers and remove listeners. */
    stop() {
        this._started = false;
        this._stopPolling();
        this._clearDebounce();
        this._clearRetry();
        window.removeEventListener('online', this._onOnline);
        window.removeEventListener('offline', this._onOffline);
        document.removeEventListener('visibilitychange', this._onVisibilityChange);
    }

    /** Called by mutation functions. Starts/resets the debounce timer. */
    scheduleUpload() {
        this._clearDebounce();
        this._debounceTimer = setTimeout(() => {
            this._debounceTimer = null;
            this._executeSync('debounce');
        }, this._uploadDebounceMs);
    }

    /** Immediate sync request. Clears pending debounce (full sync uploads dirty data). */
    requestSync() {
        this._clearDebounce();
        this._clearRetry();
        this._executeSync('request');
    }

    /** Reset retry state. Called after successful force sync. */
    resetRetry() {
        this._retryAttempt = 0;
        this._clearRetry();
    }

    // --- Internal ---

    async _executeSync(trigger) {
        if (!navigator.onLine) return;

        // If already syncing, remember we were asked so we follow up after
        if (this._getIsSyncing()) {
            this._pendingSync = true;
            return;
        }

        debugLog(`${this._name}-scheduler`, 'sync triggered', { trigger, retryAttempt: this._retryAttempt });

        try {
            const result = await this._syncFn();

            if (result.success || result.reason === 'conflicts') {
                // Success or conflicts (handled by journal UI) — reset retry
                this._retryAttempt = 0;
                this._clearRetry();
            } else if (result.reason === 'offline' || result.reason === 'already syncing') {
                // Not an error, just skip
                return;
            } else if (result.error) {
                // syncFn returned a failure with an error object
                this._handleError(result.error);
            } else {
                // Generic failure — schedule retry without toast
                this._scheduleRetry();
            }
        } catch (error) {
            this._handleError(error);
        } finally {
            // After sync completes, check if new dirty data accumulated during sync
            if (this._pendingSync || this._getHasDirtyData()) {
                this._pendingSync = false;
                this.scheduleUpload();
            }
        }
    }

    _handleError(error) {
        const errorType = this._classifyError(error);
        debugLog(`${this._name}-scheduler`, 'sync error', { errorType, message: error.message });

        if (errorType === 'server') {
            showNotification({
                type: 'error',
                title: 'Sync Failed',
                message: error.message,
                duration: 5000
            });
        }
        // Network errors: silent — indicator already pulsed during attempt
        this._scheduleRetry();
    }

    _classifyError(error) {
        // TypeError = network-level failure (DNS, offline, CORS)
        if (error instanceof TypeError) return 'network';
        if (error.name === 'AbortError') return 'network';
        if (!navigator.onLine) return 'network';
        return 'server';
    }

    _scheduleRetry() {
        this._clearRetry();
        const delay = Math.min(
            this._baseRetryMs * Math.pow(2, this._retryAttempt),
            this._maxRetryMs
        );
        this._retryAttempt++;
        debugLog(`${this._name}-scheduler`, 'retry scheduled', { delay, attempt: this._retryAttempt });
        this._retryTimer = setTimeout(() => {
            this._retryTimer = null;
            this._executeSync('retry');
        }, delay);
    }

    _startPolling() {
        if (this._pollTimer) return;
        this._pollTimer = setInterval(() => this._poll(), this._pollIntervalMs);
    }

    _stopPolling() {
        if (this._pollTimer) {
            clearInterval(this._pollTimer);
            this._pollTimer = null;
        }
    }

    async _poll() {
        if (!navigator.onLine || this._getIsSyncing()) return;

        if (this._pollCheckFn) {
            try {
                const shouldSync = await this._pollCheckFn();
                if (!shouldSync) return;
            } catch {
                // Poll check failed — skip this cycle
                return;
            }
        }

        this._executeSync('poll');
    }

    _onOnline() {
        this.requestSync();
        this._startPolling();
    }

    _onOffline() {
        this._stopPolling();
        this._clearDebounce();
        this._clearRetry();
    }

    _onVisibilityChange() {
        if (document.visibilityState === 'visible' && navigator.onLine) {
            this.requestSync();
            this._startPolling();
        } else {
            this._stopPolling();
        }
    }

    _clearDebounce() {
        if (this._debounceTimer) {
            clearTimeout(this._debounceTimer);
            this._debounceTimer = null;
        }
    }

    _clearRetry() {
        if (this._retryTimer) {
            clearTimeout(this._retryTimer);
            this._retryTimer = null;
        }
    }
}

File: android/core/data/src/main/kotlin/dev/jtiisto/wellness/core/data/sync/SyncScheduler.kt
package dev.jtiisto.wellness.core.data.sync

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Sync triggering for one module — a coroutine port of `sync-scheduler.js`.
 *
 * Debounced uploads, foreground polling, retry with backoff, and the
 * `pendingSync` re-arm, all driven by [SyncOrchestrator]. One instance per
 * module, living for the process, not for an activity.
 *
 * Two invariants carry the whole design:
 *
 * 1. **Timers and the flight are separate jobs.** The debounce, retry and poll
 *    timers are cancellable at will; a running [syncFn] never is. Cancelling a
 *    timer mid-flight would abandon an upload halfway through.
 * 2. **State is confined to the main dispatcher.** Every field below is read
 *    and written only from [control], which makes the busy check atomic: two
 *    concurrent triggers cannot both enter [syncFn]; the loser sets
 *    `pendingSync` and the winner's `finally` picks it up.
 *
 * Deviations from the PWA are declared in `specs/core-plumbing.md`: the poll
 * loop is sequential rather than fixed-rate, polling is foreground-gated, and
 * single-flight is enforced here rather than merely advised by [isSyncing].
 *
 * @param scope app-lived (survives backgrounding); tests pass a `TestScope`.
 * @param isSyncing the owning store's own notion of "busy", checked alongside
 *   this scheduler's internal flight.
 * @param isNetworkError decides silent-retry (network) vs [onServerError].
 */
class SyncScheduler(
    scope: CoroutineScope,
    private val name: String,
    private val syncFn: suspend () -> SyncResult,
    private val isSyncing: () -> Boolean,
    private val hasDirtyData: suspend () -> Boolean,
    private val isOnline: () -> Boolean,
    private val pollCheckFn: (suspend () -> Boolean)? = null,
    private val onServerError: (Throwable) -> Unit = {},
    private val isNetworkError: (Throwable) -> Boolean,
    private val uploadDebounceMs: Long = 2_500,
    private val pollIntervalMs: Long = 30_000,
    private val baseRetryMs: Long = 5_000,
    private val maxRetryMs: Long = 120_000,
    private val debugLog: DebugLog? = null,
) {
    private val control = scope + Dispatchers.Main.immediate
    private val logTag = "$name-scheduler"

    private var debounceJob: Job? = null
    private var retryJob: Job? = null
    private var pollJob: Job? = null
    private var retryAttempt = 0
    private var pendingSync = false
    private var syncInFlight = false
    private var stopped = false

    /** Called by mutators after marking data dirty. Resets the debounce timer. */
    fun scheduleUpload() {
        control.launch { armDebounce() }
    }

    /**
     * Sync now: drop the pending debounce and the pending retry *timer*, then
     * execute. The retry attempt counter survives — an immediate request is not
     * evidence that the server recovered.
     *
     * Returns the job that does it, as [stop] does and for a related reason: a
     * caller that asked for this sync on the user's behalf — the pull gesture —
     * has to be able to tell when it is over.
     *
     * **The job is not always the flight.** In the ordinary case it is: the body
     * awaits [runSync], which awaits [syncFn]. But the busy path returns as soon
     * as it has set `pendingSync` (the real flight belongs to another job), and
     * so does the offline check — so a caller waiting for "the sync finished"
     * must also wait for the owning store's own busy flag to clear. That flag is
     * observable on both stores precisely for this.
     *
     * @param trigger what to record in the debug log as the cause. Callers
     *   outside this class pass [TRIGGER_PULL]; everything else is internal.
     */
    fun requestSync(trigger: String = TRIGGER_REQUEST): Job = control.launch {
        if (stopped) return@launch
        cancelDebounce()
        cancelRetryTimer()
        runSync(trigger)
    }

    /** Clear retry state entirely, timer and counter. The force-sync hook. */
    fun resetRetry() {
        control.launch {
            if (stopped) return@launch
            retryAttempt = 0
            cancelRetryTimer()
        }
    }

    /**
     * Connectivity came back. Flushes dirty data even when backgrounded;
     * whether to also poll is the orchestrator's call.
     */
    fun onOnline() {
        requestSync()
    }

    /** Connectivity went away: stop polling and drop the timers (not the counter). */
    fun onOffline() {
        control.launch {
            cancelPolling()
            cancelDebounce()
            cancelRetryTimer()
        }
    }

    /** Begin the poll loop. Idempotent. Driven by the orchestrator. */
    fun startPolling() {
        control.launch {
            if (stopped || pollJob?.isActive == true) return@launch
            pollJob = control.launch {
                // Sequential by construction: the next wait starts only once the
                // previous cycle finished, so a slow pollCheckFn can never
                // overlap itself. Cadence drifts by the cycle duration.
                while (isActive) {
                    delay(pollIntervalMs)
                    pollOnce()
                }
            }
        }
    }

    /** Stop the poll loop. Idempotent. */
    fun stopPolling() {
        control.launch { cancelPolling() }
    }

    /**
     * Terminal and idempotent: every timer is cancelled and later calls no-op.
     * A [syncFn] already in flight runs to completion — we never leave an
     * upload half-applied.
     *
     * Returns the job that does it, because the work is dispatched to [control]
     * rather than performed here: the server switch has to *know* no new cycle
     * can be scheduled before it drains, and calling this from a background
     * thread only queues that onto the main dispatcher. Callers with no such
     * ordering requirement can ignore it.
     */
    fun stop(): Job = control.launch {
        stopped = true
        cancelPolling()
        cancelDebounce()
        cancelRetryTimer()
    }

    // --- Internal: all of the below runs on [control] ---

    /**
     * A fired timer is left in its field rather than nulled from inside the
     * coroutine: on an immediate dispatcher the body can run before the
     * assignment lands, and a self-clear would then be overwritten. Cancelling
     * an already-completed job is a no-op, so nothing depends on the difference.
     */
    private fun armDebounce() {
        if (stopped) return
        cancelDebounce()
        debounceJob = control.launch {
            delay(uploadDebounceMs)
            executeSync(TRIGGER_DEBOUNCE)
        }
    }

    /**
     * Start a sync in a job of its own. Timer coroutines call this and then
     * finish, so cancelling a timer can never reach into the flight.
     */
    private fun executeSync(trigger: String) {
        control.launch { runSync(trigger) }
    }

    private suspend fun runSync(trigger: String) {
        if (stopped) return
        if (!isOnline()) return

        // Busy check sits *before* the try/finally on purpose: this path must
        // not run the finally block. The flight that is actually running owns
        // the flag and consumes it when it completes.
        if (isSyncing() || syncInFlight) {
            pendingSync = true
            // …unless the flight is not ours. When the busy-ness comes from the
            // store's own flag — the flush Worker, the force-sync orchestrator
            // — no `finally` of this scheduler will ever run to consume the
            // latch, and the trigger would be dropped until some unrelated edit
            // came along. Arming the timer here is what that missing finally
            // would have done; it re-checks in one debounce and costs nothing
            // while the external syncer is still working.
            if (!syncInFlight) armDebounce()
            return
        }

        syncInFlight = true
        debugLog?.log(logTag, "sync triggered", buildJsonObject {
            put("trigger", trigger)
            put("retryAttempt", retryAttempt)
        })

        try {
            val result = syncFn()
            when (classifySyncOutcome(result)) {
                SyncOutcome.RESET -> {
                    retryAttempt = 0
                    cancelRetryTimer()
                }
                // Not an error, nothing to reset or retry. The finally still
                // runs, as it does for the JS `return` inside its try.
                SyncOutcome.SKIP -> return
                // ERROR is only classified when an error is present.
                SyncOutcome.ERROR -> handleError(requireNotNull(result.error))
                SyncOutcome.RETRY -> scheduleRetry()
            }
        } catch (cancellation: CancellationException) {
            // Never classified, never retried: cancellation is not a failure.
            throw cancellation
        } catch (error: Throwable) {
            handleError(error)
        } finally {
            syncInFlight = false
            if (pendingSync || hasDirtyDataQuietly()) {
                pendingSync = false
                armDebounce()
            }
        }
    }

    private fun handleError(error: Throwable) {
        val networkError = isNetworkError(error)
        // The class name, never `message`: a Ktor ResponseException carries the
        // whole response body in its message, and this log is shareable.
        debugLog?.log(logTag, "sync error", buildJsonObject {
            put("errorType", if (networkError) "network" else "server")
            put("error", error::class.simpleName ?: "error")
        })
        // Network errors stay silent — the status indicator already showed it.
        if (!networkError) onServerError(error)
        scheduleRetry()
    }

    private fun scheduleRetry() {
        if (stopped) return
        cancelRetryTimer()
        val delayMs = computeRetryDelay(retryAttempt, baseRetryMs, maxRetryMs)
        retryAttempt++
        debugLog?.log(logTag, "retry scheduled", buildJsonObject {
            put("delay", delayMs)
            put("attempt", retryAttempt)
        })
        retryJob = control.launch {
            delay(delayMs)
            executeSync(TRIGGER_RETRY)
        }
    }

    private suspend fun pollOnce() {
        if (!isOnline() || isSyncing() || syncInFlight) return

        val check = pollCheckFn
        if (check != null) {
            val shouldSync = try {
                check()
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (_: Throwable) {
                // Poll check failed — skip this cycle silently. Not a sync
                // failure: no error surfaced, no retry armed.
                return
            }
            if (!shouldSync) return
        }

        executeSync(TRIGGER_POLL)
    }

    /**
     * A dirty-data probe that throws would escape the `finally` and take the
     * whole app scope with it. Failing closed just delays the next upload to
     * the following trigger.
     */
    private suspend fun hasDirtyDataQuietly(): Boolean = try {
        hasDirtyData()
    } catch (_: Throwable) {
        false
    }

    private fun cancelDebounce() {
        debounceJob?.cancel()
        debounceJob = null
    }

    private fun cancelRetryTimer() {
        retryJob?.cancel()
        retryJob = null
    }

    private fun cancelPolling() {
        pollJob?.cancel()
        pollJob = null
    }

    companion object {
        /**
         * The pull gesture, and the only trigger name with a caller outside this
         * class. The other three describe timers nobody else can start, so they
         * stay private rather than widening the public surface for symmetry.
         */
        const val TRIGGER_PULL = "pull"

        private const val TRIGGER_DEBOUNCE = "debounce"
        private const val TRIGGER_REQUEST = "request"
        private const val TRIGGER_RETRY = "retry"
        private const val TRIGGER_POLL = "poll"
    }
}
